shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXConnection.scala

```diff
@@ -40,21 +45,98 @@ private[ucx] abstract class UCXTagCallback {
   def onCancel(alt: AddressLengthTag): Unit
 }
 
-class UCXServerConnection(ucx: UCX) extends UCXConnection(ucx) with ServerConnection {
+/**
+ * `UCXAmCallback` is used by [[Transaction]] to handle UCX Active Messages operations.
+ * The `UCXActiveMessage` object encapsulates an activeMessageId and a header.
+ */
+private[ucx] abstract class UCXAmCallback {
+  def onError(am: UCXActiveMessage, ucsStatus: Int, errorMsg: String): Unit
+  def onMessageStarted(receiveAm: UcpRequest): Unit
+  def onSuccess(am: UCXActiveMessage, buff: RefCountedDirectByteBuffer): Unit
+  def onCancel(am: UCXActiveMessage): Unit
+
+  // hook to allocate memory on the host
+  // a similar hook will be needed for GPU memory
+  def onHostMessageReceived(size: Long): RefCountedDirectByteBuffer
+}
+
+class UCXServerConnection(ucx: UCX, transport: UCXShuffleTransport)
+  extends UCXConnection(ucx) with ServerConnection with Logging {
   override def startManagementPort(host: String): Int = {
     ucx.startManagementPort(host)
   }
 
-  override def send(sendPeerExecutorId: Long, bounceBuffers: Seq[AddressLengthTag],
+  override def send(sendPeerExecutorId: Long, buffer: AddressLengthTag,
     cb: TransactionCallback): Transaction =
-    send(sendPeerExecutorId, null, bounceBuffers, cb)
+    send(sendPeerExecutorId, buffer, Seq.empty, cb)
 
-  override def send(sendPeerExecutorId: Long, header: AddressLengthTag,
-    cb: TransactionCallback): Transaction =
-    send(sendPeerExecutorId, header, Seq.empty, cb)
+  override def registerRequestHandler(
+      requestType: RequestType.Value, cb: TransactionCallback): Unit = {
+
+    ucx.registerRequestHandler(
+        UCXConnection.composeRequestAmId(requestType), () => new UCXAmCallback {
+      private val tx = createTransaction
+
+      tx.start(UCXTransactionType.Request, 1, cb)
+
+      override def onSuccess(am: UCXActiveMessage, buff: RefCountedDirectByteBuffer): Unit = {
+        logDebug(s"At requestHandler for ${requestType} and am: " +
+          s"$am")
+        tx.completeWithSuccess(requestType, Option(am.header), Option(buff))
+      }
+
+      override def onHostMessageReceived(size: Long): RefCountedDirectByteBuffer = {
```

Comment:
I would consider initially handle both metadata and data buffers with AM. Maybe split to some public API, that would handle both cases. Even left `onDeviceMessageReceived` unimplemented, but to see the whole picture. Trying to falling my or with AM: https://github.com/NVIDIA/spark-rapids/pull/2416 - want it to work with both GPU and host memory

Reply:
Should RNDV for GPU buffers work in 1.10 for Active Messages? Just checking, I have done mostly host buffers.

I would really like to add GPU support as an different PR, need to discuss with others as well on the timing.